### PR TITLE
Scripts/Spells: Stealth, Subterfuge and Subterfuge effect

### PR DESCRIPTION
**Pull request by Daniel25**
-  Stealh working, it wasnt working.
-  Subterfuge working, it wasnt working.

- [x] master
- last commit: https://github.com/TrinityCore/TrinityCore/commit/52c7f12ed35f321dc97d8caea2b112287f5ac05b

**It builds and works in game.

### CODE REVIEW

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 1]
```diff
@@ -602,65 +607,85 @@ class spell_rog_shiv : public SpellScriptLoader
...
+                OnEffectApply += AuraEffectApplyFn(spell_rog_stealth_AuraScript::OnApply, EFFECT_0, SPELL_AURA_DUMMY, AURA_EFFECT_HANDLE_REAL);
+                OnEffectRemove += AuraEffectRemoveFn(spell_rog_stealth_AuraScript::OnRemove, EFFECT_0, SPELL_AURA_DUMMY, AURA_EFFECT_HANDLE_REAL);
             }
         };
 
-        AuraScript* GetAuraScript() const override
-        {
-            return new spell_rog_stealth_AuraScript();
-        }
+    AuraScript* GetAuraScript() const
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: bad indentation, bad override removal

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 2]
```diff
@@ -602,65 +607,85 @@ class spell_rog_shiv : public SpellScriptLoader
...
+            StealthSubterfugeEffect = 115192,
+            StealthTriggered1 = 158188,
+            StealthTriggered2 = 158185
+        };
+
         class spell_rog_stealth_AuraScript : public AuraScript
         {
             PrepareAuraScript(spell_rog_stealth_AuraScript);
 
-            bool Validate(SpellInfo const* /*spellInfo*/) override
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: it is bad to remove this validate, you need to update the spells

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 3]
```diff
@@ -602,65 +607,85 @@ class spell_rog_shiv : public SpellScriptLoader
         }
 };
 
-// 1784 - Stealth
+/// Stealth - 1784, Subterfuge - 115191 and Subterfuge effect - 115192
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: don't change this comment if you not bind the spellscript to all these 3 spells

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 4]
```diff
@@ -32,6 +32,7 @@
 #include "SpellHistory.h"
 #include "SpellMgr.h"
 #include "SpellScript.h"
+#include "GridNotifiers.h"
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: why?

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 5]
```diff
@@ -602,65 +607,85 @@ class spell_rog_shiv : public SpellScriptLoader
...
 
-                target->CastSpell(target, SPELL_ROGUE_SANCTUARY, TRIGGERED_FULL_MASK);
-                target->CastSpell(target, SPELL_ROGUE_STEALTH_STEALTH_AURA, TRIGGERED_FULL_MASK);
-                target->CastSpell(target, SPELL_ROGUE_STEALTH_SHAPESHIFT_AURA, TRIGGERED_FULL_MASK);
+                        if (l_Caster->HasAura(SPELL_ROGUE_SHADOW_FOCUS_AURA))
+                            l_Caster->CastSpell(l_Caster, SPELL_ROGUE_SHADOW_FOCUS_COST_PCT, true);
+                    }
+                    else
+                    {
+                        if (l_Caster->HasAura(eSpells::StealthSubterfuge))
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: useless hasaura check

**Comment by HannibalRoG** [comment]
This PR close https://github.com/TrinityCore/TrinityCore/issues/19712

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 6]
```diff
@@ -640,15 +652,11 @@ class spell_rog_stealth : public SpellScriptLoader
                         if (l_Caster->HasAura(SPELL_ROGUE_SHADOW_FOCUS_AURA))
                             l_Caster->CastSpell(l_Caster, SPELL_ROGUE_SHADOW_FOCUS_COST_PCT, true);
                     }
-                    else
-                    {
-                        if (l_Caster->HasAura(eSpells::StealthSubterfuge))
-                            l_Caster->RemoveAura(eSpells::StealthSubterfuge);
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: why do you remove the `RemoveAura` now?

Comment (Daniel25) [reply src/server/scripts/Spells/spell_rogue.cpp]: Sorry Im doing too much things in the same time...

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 7]
```diff
@@ -622,38 +635,61 @@ class spell_rog_stealth : public SpellScriptLoader
...
             {
-                Unit* target = GetTarget();
+                if (Unit* l_Caster = GetCaster())
+                {
+                    if (GetSpellInfo()->Id != eSpells::StealthSubterfugeEffect)
+                    {
+                        l_Caster->CastSpell(l_Caster, eSpells::StealthTriggered1, true);
+                        l_Caster->CastSpell(l_Caster, eSpells::StealthTriggered2, true);
 
-                // Master of Subtlety
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: what happend with Master of Subtlety?

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 8]
```diff
@@ -621,33 +634,72 @@ class spell_rog_stealth : public SpellScriptLoader
...
+                    SPELL_ROGUE_NIGHTSTALKER_DAMAGE_DONE,
+                    SPELL_ROGUE_SHADOW_FOCUS_AURA,
+                    SPELL_ROGUE_SHADOW_FOCUS_COST_PCT
                 });
             }
 
             void HandleEffectApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
             {
-                Unit* target = GetTarget();
+                if (Unit* target = GetTarget())
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: target never null

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 9]
```diff
@@ -657,10 +709,10 @@ class spell_rog_stealth : public SpellScriptLoader
             }
         };
 
-        AuraScript* GetAuraScript() const override
-        {
-            return new spell_rog_stealth_AuraScript();
-        }
+		AuraScript* GetAuraScript() const
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tabs2spaces.

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 10]
```diff
@@ -642,11 +642,11 @@ class spell_rog_stealth : public SpellScriptLoader
                 });
             }
 
-            void HandleEffectApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+	    void HandleEffectApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tab2spaces.

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 11]
```diff
@@ -607,6 +611,15 @@ class spell_rog_stealth : public SpellScriptLoader
 {
     public:
         spell_rog_stealth() : SpellScriptLoader("spell_rog_stealth") { }
+
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: remove whitespaces.

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 12]
```diff
@@ -621,33 +634,72 @@ class spell_rog_stealth : public SpellScriptLoader
...
+		target->CastSpell(target, SPELL_ROGUE_SANCTUARY, TRIGGERED_FULL_MASK);
+		target->CastSpell(target, SPELL_ROGUE_STEALTH_STEALTH_AURA, TRIGGERED_FULL_MASK);
+		target->CastSpell(target, SPELL_ROGUE_STEALTH_SHAPESHIFT_AURA, TRIGGERED_FULL_MASK);
+	    }
 
             void HandleEffectRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
             {
-                Unit* target = GetTarget();
+	    	Unit* target = GetTarget();
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tab2spaces.

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 13]
```diff
@@ -621,33 +634,72 @@ class spell_rog_stealth : public SpellScriptLoader
...
-                    target->CastSpell(target, SPELL_ROGUE_MASTER_OF_SUBTLETY_PERIODIC, true);
+                    target->RemoveAurasDueToSpell(eSpells::StealthTriggered2);
+		}	
+		   // Master of subtlety
+		   if (target->HasAura(SPELL_ROGUE_MASTER_OF_SUBTLETY_PASSIVE))
+			target->CastSpell(target, SPELL_ROGUE_MASTER_OF_SUBTLETY_PERIODIC, true);
 
-                target->RemoveAurasDueToSpell(SPELL_ROGUE_STEALTH_STEALTH_AURA);
-                target->RemoveAurasDueToSpell(SPELL_ROGUE_STEALTH_SHAPESHIFT_AURA);
+		   target->RemoveAurasDueToSpell(SPELL_ROGUE_STEALTH_STEALTH_AURA);
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tab2spaces

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 14]
```diff
@@ -657,10 +709,10 @@ class spell_rog_stealth : public SpellScriptLoader
             }
         };
 
-        AuraScript* GetAuraScript() const override
-        {
-            return new spell_rog_stealth_AuraScript();
-        }
+	AuraScript* GetAuraScript() const
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tab2spaces

Comment (Daniel25) [reply src/server/scripts/Spells/spell_rogue.cpp]: Sorry dunno what is going on with whitespaces. Its ok in my screen but its broken then... Does someone care in editing that mistakes please?

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 15]
```diff
@@ -605,9 +609,18 @@ class spell_rog_shiv : public SpellScriptLoader
 // 1784 - Stealth
 class spell_rog_stealth : public SpellScriptLoader
 {
-    public:
+     public:
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: remove aditional whitespaces, we use 4 spaces, not 5.

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 16]
```diff
@@ -605,9 +609,18 @@ class spell_rog_shiv : public SpellScriptLoader
...
         spell_rog_stealth() : SpellScriptLoader("spell_rog_stealth") { }
 
+       	enum eSpells
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tab2spaces

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 17]
```diff
@@ -621,33 +634,72 @@ class spell_rog_stealth : public SpellScriptLoader
...
-
-                // Master of Subtlety
-                if (AuraEffect const* aurEff = target->GetAuraEffect(SPELL_ROGUE_MASTER_OF_SUBTLETY_PASSIVE, EFFECT_0))
-                    target->CastCustomSpell(SPELL_ROGUE_MASTER_OF_SUBTLETY_DAMAGE_PERCENT, SPELLVALUE_BASE_POINT0, aurEff->GetAmount(), target, true);
-
-                target->CastSpell(target, SPELL_ROGUE_SANCTUARY, TRIGGERED_FULL_MASK);
-                target->CastSpell(target, SPELL_ROGUE_STEALTH_STEALTH_AURA, TRIGGERED_FULL_MASK);
-                target->CastSpell(target, SPELL_ROGUE_STEALTH_SHAPESHIFT_AURA, TRIGGERED_FULL_MASK);
-            }
+	    	Unit* target = GetTarget();
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tabs2spaces

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 18]
```diff
@@ -621,33 +634,72 @@ class spell_rog_stealth : public SpellScriptLoader
...
+
+		target->CastSpell(target, SPELL_ROGUE_SANCTUARY, TRIGGERED_FULL_MASK);
+		target->CastSpell(target, SPELL_ROGUE_STEALTH_STEALTH_AURA, TRIGGERED_FULL_MASK);
+		target->CastSpell(target, SPELL_ROGUE_STEALTH_SHAPESHIFT_AURA, TRIGGERED_FULL_MASK);
+	    }
 
             void HandleEffectRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
             {
-                Unit* target = GetTarget();
+		Unit* target = GetTarget();
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tab2spaces

**Comment by joschiwald** [comment]
stop doing random commits, it looks like you don't know what you are doing

just replace your tabs with 4 spaces

**Comment by Daniel25** [comment]
Well i said it before....dunno what is going on with my screen but its ok and when i edited it, it isnt....

**Comment by joschiwald** [comment]
exactly, you insert spaces and tabs randomly to looks fine on your screen, you need to know what you are inserting, don't use the github editor, use some editor like notepad++ and enable display of tabs and spaces

**Comment by Daniel25** [comment]
I was using notepadd++ and when I copy it in the github editor its giving the troubles....sorry

**Comment by Daniel25** [comment]
Do you care in editing it please? We will finish before if someone can do it in my place....

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 19]
```diff
@@ -608,6 +612,15 @@ class spell_rog_stealth : public SpellScriptLoader
     public:
         spell_rog_stealth() : SpellScriptLoader("spell_rog_stealth") { }
 
+       	enum eSpells
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tabs2spaces

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 20]
```diff
@@ -621,34 +634,73 @@ class spell_rog_stealth : public SpellScriptLoader
...
+                    SPELL_ROGUE_STEALTH_SHAPESHIFT_AURA,
+                    SPELL_ROGUE_NIGHTSTALKER_AURA,
+                    SPELL_ROGUE_NIGHTSTALKER_DAMAGE_DONE,
+                    SPELL_ROGUE_SHADOW_FOCUS_AURA,
+                    SPELL_ROGUE_SHADOW_FOCUS_COST_PCT
                 });
             }
 
-            void HandleEffectApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+	  void HandleEffectApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tabs2spaces

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 21]
```diff
@@ -621,34 +634,73 @@ class spell_rog_stealth : public SpellScriptLoader
...
+
+		target->CastSpell(target, SPELL_ROGUE_SANCTUARY, TRIGGERED_FULL_MASK);
+		target->CastSpell(target, SPELL_ROGUE_STEALTH_STEALTH_AURA, TRIGGERED_FULL_MASK);
+		target->CastSpell(target, SPELL_ROGUE_STEALTH_SHAPESHIFT_AURA, TRIGGERED_FULL_MASK);
+	    }
+
+	    void HandleEffectRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
             {
-                Unit* target = GetTarget();
+		Unit* target = GetTarget();
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tabs2spaces

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 22]
```diff
@@ -621,34 +634,73 @@ class spell_rog_stealth : public SpellScriptLoader
...
+                    target->RemoveAurasDueToSpell(eSpells::StealthTriggered2);
+		}	
+		// Master of subtlety
+		if (target->HasAura(SPELL_ROGUE_MASTER_OF_SUBTLETY_PASSIVE))
+			target->CastSpell(target, SPELL_ROGUE_MASTER_OF_SUBTLETY_PERIODIC, true);
 
-                target->RemoveAurasDueToSpell(SPELL_ROGUE_STEALTH_STEALTH_AURA);
-                target->RemoveAurasDueToSpell(SPELL_ROGUE_STEALTH_SHAPESHIFT_AURA);
-            }
+		target->RemoveAurasDueToSpell(SPELL_ROGUE_STEALTH_STEALTH_AURA);
```
Comment (Aokromes) [inline src/server/scripts/Spells/spell_rogue.cpp]: tabs2spaces

**Comment by Daniel25** [comment]
It isnt working now...dont merge it still....

**Comment by Daniel25** [comment]
With all that changes the fix doesnt work now....So dont merge it still...I have to check it again....

**Comment by Daniel25** [comment]
Validate funtion is commented because the fix doesnt work with the funtion active. I guess that its what broken the stealth spell before.
It compiles and works good now.

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 23]
```diff
@@ -605,62 +611,105 @@ class spell_rog_shiv : public SpellScriptLoader
 // 1784 - Stealth
 class spell_rog_stealth : public SpellScriptLoader
 {
-    public:
-        spell_rog_stealth() : SpellScriptLoader("spell_rog_stealth") { }
+public:
+    spell_rog_stealth() : SpellScriptLoader("spell_rog_stealth") { }
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: pls restore the original indentation

**Comment by Daniel25** [comment]
Thanks you for the patiente and sorry for the mistakes. Its my first pull request :/.

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 24]
```diff
@@ -612,23 +618,46 @@ class spell_rog_stealth : public SpellScriptLoader
...
+          //        SPELL_ROGUE_SHADOW_FOCUS_COST_PCT,
+          //        SPELL_ROGUE_STEALTH_SUBTERFUGE,
+          //        SPELL_ROGUE_STEALTH_SUBTERFUGE_EFFECT
+          //     });
+          //}
 
             void HandleEffectApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
             {
                 Unit* target = GetTarget();
+                if (Unit* target = GetTarget())
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: useless if condition

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 25]
```diff
@@ -641,7 +670,27 @@ class spell_rog_stealth : public SpellScriptLoader
             void HandleEffectRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
             {
                 Unit* target = GetTarget();
+                if (Unit* target = GetTarget())
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: useless if condition

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 26]
```diff
@@ -657,7 +706,7 @@ class spell_rog_stealth : public SpellScriptLoader
             }
         };
 
-        AuraScript* GetAuraScript() const override
+        AuraScript* GetAuraScript() const
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: you removed override keyword again

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 27]
```diff
@@ -641,7 +670,27 @@ class spell_rog_stealth : public SpellScriptLoader
...
             {
                 Unit* target = GetTarget();
+                if (Unit* target = GetTarget())
+                {
+                    AuraRemoveMode removeMode = GetTargetApplication()->GetRemoveMode();
+                    if (GetSpellInfo()->Id != SPELL_ROGUE_STEALTH_SUBTERFUGE || removeMode == AURA_REMOVE_BY_CANCEL)
+                    {
+                        target->RemoveAura(SPELL_ROGUE_STEALTH_SHAPESHIFT_AURA);
+
+                        if (Aura* l_Nightstalker = target->GetAura(SPELL_ROGUE_NIGHTSTALKER_DAMAGE_DONE))
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: `nightstalker`

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 28]
```diff
@@ -641,7 +670,27 @@ class spell_rog_stealth : public SpellScriptLoader
...
+                            l_Nightstalker->SetDuration(200);   ///< We can't remove it now
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: why?

File: src/server/scripts/Spells/spell_rogue.cpp [hunk 29]
```diff
@@ -612,23 +618,46 @@ class spell_rog_stealth : public SpellScriptLoader
...
+          //        SPELL_ROGUE_STEALTH_SUBTERFUGE_EFFECT
+          //     });
+          //}
 
             void HandleEffectApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
             {
                 Unit* target = GetTarget();
+                if (Unit* target = GetTarget())
+                {
+                    if (GetSpellInfo()->Id != SPELL_ROGUE_STEALTH_SUBTERFUGE_EFFECT)
```
Comment (joschiwald) [inline src/server/scripts/Spells/spell_rogue.cpp]: this will never work, because the spell is not bound to this script

Comment (Daniel25) [reply src/server/scripts/Spells/spell_rogue.cpp]: Sorry I dont know to bound it.

**Comment by Daniel25** [comment]
If I revert SPELL_ROGUE_STEALTH_SUBTERFUGE and SPELL_ROGUE_STEALTH_SUBTERFUGE_EFFECT to what I had in the begin,will it work?

**Comment by joschiwald** [comment]
broken code

supercedes by b12b37058c863bcfecfefc965da45f31b604ee3b

**Pull request closed without merging**